libs/mpi/tests/unit/mpi_ring_async_executor.cpp

```diff
@@ -74,63 +89,84 @@ int hpx_main(hpx::program_options::variables_map& vm)
 
         hpx::mpi::experimental::executor exec(MPI_COMM_WORLD);
 
-        constexpr unsigned int n_loops = 20;
-        std::atomic<int> counter(n_loops);
-        std::array<int, n_loops> tokens;
-        for (unsigned int i = 0; i != n_loops; ++i)
+        // mpi chokes if we put too many messages into the system at once
+        // so do iterations in batches of 1000
+        // using an inner and outer loop
+        const std::uint64_t max_tokens = 1000;
+        std::uint64_t subloops  = 1 + iterations/max_tokens;
+        std::uint64_t remainder = iterations;
+
+        std::vector<int> tokens(max_tokens, -1);
+
+        hpx::util::high_resolution_timer t;
+
+        for (std::uint64_t j = 0; (j!=subloops) && (remainder>0); ++j)
         {
-            tokens[i] = (rank == 0) ? 1 : -1;
-            int rank_from = (size + rank - 1) % size;
-            int rank_to = (rank + 1) % size;
-
-            // all ranks pre-post a receive
-            hpx::future<int> f_recv = hpx::async(
-                exec, MPI_Irecv, &tokens[i], 1, MPI_INT, rank_from, i);
-
-            // when the recv completes,
-            f_recv.then([=, &tokens, &counter](auto&&) {
-                msg_recv(rank, size, rank_to, rank_from, tokens[i], i);
-                if (rank > 0)
+            std::atomic<int> counter((std::min)(std::uint64_t(max_tokens), remainder));
+            for (std::uint64_t i = 0; (i!=max_tokens) && (remainder>0); ++i)
+            {
+                tokens[i] = (rank == 0) ? 1 : -1;
+                int rank_from = (size + rank - 1) % size;
+                int rank_to = (rank + 1) % size;
+
+                // all ranks pre-post a receive
+                hpx::future<int> f_recv = hpx::async(
+                    exec, MPI_Irecv, &tokens[i], 1, MPI_INT, rank_from, i);
+
+                // when the recv completes,
+                f_recv.then([=, &tokens, &counter](auto&&) {
```

Comment:
Discarding the returned future counters all of our paradigms. Should we adapt this test to follow the usual 'protocol'?

Reply:
what do you mean by the 'usual' protocol?

Reply:
I would consider not discarding any futures returned from our APIs being good practice.

Reply:
@biddisco what's the plan here now? Do we keep things as they are or are you planning to change this still?

Reply:
I tried an experiment where I collected all the futures and waited for them properly instead of using the atomic counter, but it ran about 5% more slowly than the existing version. I would prefer to leave this as is for now and get the PR merged, then if someone wants to clean up the future handling at some point they can. 
If we had an `apply` version of `.then`, we would not need to drop any but could instead decrement counters in continuations that don't return a future.
```
f1 = hpx::async(some_work);
f2 = f1.then(other_work); // ok
// we don't have
f2.then_apply(no_future_returned_from_this);
```